design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+* Multiple table functions on the SELECT clause
+* A set of built-in table functions such as EXPLODE (etc)
+* An ability for users to provide their own User Defined Table Functions (UDTFs) using annotations
+in a similar way to how UDFs and UDAFs are currently supported.
+
+## What is not in scope
+
+* Table functions on the FROM clause. We could allow table functions to take the place of other streams
+or tables in the FROM clause. We do not think there is much value in doing this as the main driver
+for this KLIP is in supporting `explode` functionality which is more simply implemented on the
+SELECT clause.
+* Table functions that output more than one column. Some databases support the flattening of an array
+ of N STRUCT types into N rows of M columns, where there is one column for each field of the struct.
+ Instead we propose that flattening an array of N structs will result in N rows with a single column
+ of type STRUCT. The fields of the struct can then be extracted if necessary using a further CSAS
+ which pulls out the struct fields.
+
+## Value/Return
+
+This functionality will make it easier for users to make end-end streaming apps using KSQL
+alone without having to have custom stages using Kafka Streams.
+
+We believe its a common case for users to have array or map like data structures as values
+in messages that they wish to explode. This new feature would allow them to do that in a pure KSQL
+app thus allowing them to create richer KSQL applications.
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDAFs.
+
+## Design
+
+### General plumbing
+
+We will need the following general plumbing to support table functions:
+
+* In a similar way to how UDAFs are currently supported we will need to create a 
+`KsqlTableFunction` which is a wrapper around the actual table function (whether built-in or
+generated). We'll also need to augment the function registry to manage the addition and retrieval
+of table functions.
+* Augment the query analysis code to extract information relating to any table functions
+in the query. We can do this either by extending the existing query analyzer or perhaps creating
+a new `TableFunctionAnalyzer` in a similar way to how aggregate analysis uses its own 
+`AggregateAnalyzer`.
+* Augment the building of logical plan by building a new node representing a FlatMap stage of the
+stream processing in the case there are table functions in the SELECT.
+* Augment the execution plan logic by adding new nodes in the execution plan corresponding to any
+extra FlatMap stage required by the query.
+* Amend the columns used in the final projection to reflect any generated columns created by a 
+FlatMap stage.
+* Implement the actual FlatMap logic. For each row, any table functions on that row will be evaluated.
```

Comment:
> The number of rows returned from the FlatMap will be equal to the greatest of the number of rows returned from any of the table functions for that row.

As mentioned above, I think it should be the cartesian product.

Reply:
I discussed this quite a bit with @derekjn and others and we could think of many more useful use cases where a zipjoin was desired, (e.g. the sensors example in the KLIP) but we couldn't come up with any for a cartesian join. If you can provide an example that would help me understand the utility of the suggestion a bit better :)